app/controllers/devise/two_factor_authentication_controller.rb

```diff
@@ -86,25 +99,21 @@ def redirect_valid_resource
       redirect_to after_sign_in_path_for(resource)
     end
 
-    def show_direct_otp_prompt
-      redirect_to otp_confirm_path(otp_method: current_otp_method)
-    end
-
-    def show_totp_prompt
-      render :confirm_totp
-    end
-
     def handle_invalid_otp
       analytics.track_event('User entered invalid 2FA code')
 
       update_invalid_resource if resource.two_factor_enabled?
 
-      flash[:error] = t('devise.two_factor_authentication.attempt_failed')
+      flash[:error] = if current_otp_method == :rescue_code
+                        t('devise.two_factor_authentication.invalid_rescue_code')
+                      else
+                        t('devise.two_factor_authentication.attempt_failed')
+                      end
```

Comment:
Here's another conditional that can be eliminated by moving rescue code verification to its own controller. 

As Sandi Metz said in *Practical Object-Oriented Design in Ruby*:
> The code you write should have the following qualities. Code should be
- **Transparent** The consequences of change should be obvious in the code that is changing and in distant code that relies upon it
- **Reasonable** The cost of any change should be proportional to the benefits the change achieves
- **Usable** Existing code should be usable in new and unexpected contexts
- **Exemplary** The code itself should encourage those who change it to perpetuate these qualities 

> Code that is Transparent, Reasonable, Usable, and Exemplary (TRUE) not only meets today's needs but can also be changed to meet the needs of the future.

By moving this logic to its own controller, it lets future developers know that to add a new 2FA method in the future, it should be done via a new controller. This makes the code exemplary because adding a new option can be done by adding a new controller without modifying existing controllers (Open for Extension, Closed for Modification). If we left the logic in the same controller using params and conditionals, it would send a message to developers that it's OK to keep doing that, and we don't want that.